### DISCUSSION TITLE
Having a difficult time implementing a custom mode with custom behaviours.

### DISCUSSION THREAD

**Discussion by PhonicUK** [ajaxorg/ace#5157] [Q&A]
I'm finding the documentation overly dense and lacking practical, real world examples. The code for existing modes is also entirely uncommented.

I have a file format with the extension of ".kvp" that looks like this:

    Section.Setting=Some Text Value
    Section.Foo=123
    Section.SomeRegex=^.+?$
    AnotherSection.ComplexValue={"foo":"bar"}

Complex values include arrays, dictionaries, etc and are represented as single-line JSON. They can be arbitrarily complex with nested structures etc.

Where ComplexValues are single-line JSON blobs, what I actually want to render is show just the first few words followed by a button that says "[Edit Object]" or something to that effect, and it'll expand *in place* to a pretty-printed version of the JSON in a box (complete with normal JSON syntax highlighting and rule checking). When I retrieve the value of the document again though, it should be a single line.

Can I get some detailed guidance about how I'd go about doing this? Or frankly, if you feel you could implement this yourself - I'd be happy to simply pay a fair price for someone to implement this for me.

**Comment by nightwing** [comment]
This is possible to do, but wouldn't it be better to convert document to a form with formatted json statements when passing it to editor, use normal folding to bring multiline json into one line, and then convert it back when saving the modified document?

**Reply by PhonicUK** [reply]
The reason for this approach is that most of the time the person editing the document won't be interested in the JSON segments (they can be very large) so the idea is that they remain collapsed until they want to edit them. I feel that having it expand in place then collapse when they're done would provide a good UX because most of the time the document retains it's standard 'appearance' until you do something different by editing these fields.

**Reply by nightwing** [reply]
With folding it behaves the same way too, i guess the difference will be with ctrl-f finding text in json segments. I guess something like this may be useful for data urls in css as well. You can write me via email in my profile for a quick chat about this, if you want.
